Mathlib/Algebra/FreeMonoid/Basic.lean

```diff
@@ -415,4 +423,37 @@ theorem length_reverse {a : FreeMonoid α} : a.reverse.length = a.length :=
 
 end Reverse
 
+section IsomorphicTypes
+
+variable {α β : Type*}
+
+/-- free monoids over isomorphic types are isomorphic -/
+@[to_additive "if two types are isomorphic, the additive free monoids over those types are
+isomorphic"]
+def congrEquiv (e : α ≃ β) : FreeMonoid α ≃* FreeMonoid β :=
+  MulEquiv.mk' ⟨FreeMonoid.map e.toFun, FreeMonoid.map e.invFun, fun _ => map_symm_apply_map_eq e,
+    fun _ => map_apply_map_symm_eq e⟩ (by simp [map_mul])
+
+@[to_additive (attr := simp)]
+theorem congrEquiv_of (e : α ≃ β) (a : α) : congrEquiv e (of a) = of (e a) := rfl
+
+@[to_additive (attr := simp)]
+theorem congrEquiv_symm_of (e : α ≃ β) (b : β) : congrEquiv e.symm (of b) = of (e.symm b) := rfl
+
+/-- given a function from β to α, convert a relation predicate on FreeMonoid α to
+have an underlying type of β -/
+@[to_additive "given a function from β to α, convert a relation predicate to
+have an underlying type of β"]
+def mapRel (f : β → α) (rel : FreeMonoid α → FreeMonoid α → Prop) : FreeMonoid β → FreeMonoid β  →
```

Comment:
`mapRel` and `comapRel` are now identical (up to renaming of arguments).
My suspicion is that you will only need `comapRel`. But if you think it is useful, you can restore the version of `mapRel` that takes an equiv as argument.

Reply:
This is just `rel on (map f)`. Do we need a separate `def` for it?

Reply:
Oh nice, I didn't know about that! I've removed this definition; I can use `rel on (map f)` for my purposes then